### PR TITLE
[lodash] Add types for lodash/fp

### LINKED ISSUE #7896: Definition request: lodash/fp
In Lodash 4.X, they introduced core support for a more functional version of the library.
By importing [`'lodash/fp'`](https://github.com/lodash/lodash/tree/master/fp), you get access "to an immutable, auto-curried, iteratee-first, data-last version."

The conversion is done automatically, and basically just uses their own [`rearg`](https://lodash.com/docs#rearg) method to move around the arguments in a [standard way](https://github.com/lodash/lodash/blob/master/fp/_mapping.js#L119)*.

I'm pretty new to writing Typescript defintions but I hope that we wouldn't need a full-blown, modified version of the existing [`lodash.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/lodash/lodash.d.ts) and that we could just describe the conversion function and 'transform' the existing definition, if that's even possible.

\* It looks like there are a few [exceptions](https://github.com/lodash/lodash/blob/master/fp/_mapping.js#L133), but I don't think they'd be too problematic.

### LINKED ISSUE #9589: Lodash FP params aren't typed properly
- [x] I tried using the latest `lodash/lodash.d.ts` file in this repo and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] I want to talk about `lodash/lodash.d.ts`.
  - The authors of that type definition are cc/ @bczengel @stepancar @chrootsu 

I'm using lodash v4.13.1 and I'm noticing the fp module just points to `_`. This is no longer the case if it was in the past as the param order is swapped for some things. I would gladly make a PR if someone can point me in the right direction. 

Currently, it will show something like `get('prop', item)` as being in the wrong order but that is not the case if you [try it out](https://tonicdev.com/57095f6fd2703c1100cc5b4f/5757a90193e6131300920250) for yourself. 

Could someone tell me what to do to submit a PR or help get this updated!

### PR DESCRIPTION
**Pull request by aj-r**
Closes #7896. Closes #9589. This replaces #22587.

FP reference: https://github.com/lodash/lodash/wiki/FP-Guide

This PR adds a script that generates types for `lodash/fp`. The intention is for the script to be re-run any time a function definition changes.

`lodash/fp` can be imported in 2 ways:

```ts
import _ from "lodash/fp";
import some from "lodash/fp/some";
```

In general, I tried to avoid making changes to the original types. The only notable changes are:
- `flatten` no longer has an `isDeep` parameter. This parameter was removed in v4 and should only exist in v3.
- `indexOf` no longer accepts a boolean value for `fromIndex`. This is only allowed in v3, not v4.

The types generated by this script aren't 100% perfect, but they're pretty good. I'm hoping to sumbit this now just so people can start using it, and we can make improvements later.

Limitations (that I think we can live with for now):

- No placeholder support (e.g. `trimChars(_, "x")`)
- Some joinable overloads exist
- Generic types are not used optimally - sometimes you need to explicitly pass them when currying
- Some comments incorrectly indicate that the function will mutate the array/object. These comments are false for `fp`, because it modifies all functions to treat inputs as immutable.

Template:

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/lodash/lodash/wiki/FP-Guide
- [ ] Increase the version number in the header if appropriate. N/A
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by aj-r** [comment]
Would also aprreciate a review from @thalesmello

**Comment by aj-r** [comment]
@andy-ms Travis keeps complaining that I have a test file that isn't named `lodash-tests.d.ts`. Is there any way I can have 2 separate test files? Or do I need to throw them all into one gargantuan file?

File: types/lodash/array/flatten.d.ts [hunk 1]
```diff
@@ -2,42 +2,25 @@ import _ = require("../index");
...
         /**
-         * Flattens a nested array. If isDeep is true the array is recursively flattened, otherwise it’s only
-         * flattened a single level.
+         * Flattens `array` a single level deep.
          *
          * @param array The array to flatten.
-         * @param isDeep Specify a deep flatten.
          * @return Returns the new flattened array.
          */
-        flatten<T>(array: ListOfRecursiveArraysOrValues<T> | null | undefined, isDeep: boolean): T[];
```
Comment (aj-r) [inline types/lodash/array/flatten.d.ts]: Note: change to flatten

File: types/lodash/common/common.d.ts [hunk 2]
```diff
@@ -201,17 +201,23 @@ declare module "../index" {
...
 
+    /** @deprecated Use MemoVoidArrayIterator or MemoVoidDictionaryIterator instead. */
     type MemoVoidIterator<T, TResult> = (prev: TResult, curr: T, indexOrKey: any, list: T[]) => void;
 
-    /** @deprecated Use MemoListIterator or MemoObjectIterator instead.  */
+    /** @deprecated Use MemoListIterator or MemoObjectIterator instead. */
     type MemoIterator<T, TResult> = (prev: TResult, curr: T, indexOrKey: any, list: T[]) => TResult;
     type MemoListIterator<T, TResult, TList> = (prev: TResult, curr: T, index: number, list: TList) => TResult;
     type MemoObjectIterator<T, TResult, TList> = (prev: TResult, curr: T, key: string, list: TList) => TResult;
+    type MemoIteratorCapped<T, TResult> = (prev: TResult, curr: T) => TResult;
```
Comment (aj-r) [inline types/lodash/common/common.d.ts]: Note: added some new common types that will be shared between some `fp` functions

File: types/lodash/tsconfig.json [hunk 3]
```diff
@@ -17,306 +17,12 @@
         "forceConsistentCasingInFileNames": true,
         "esModuleInterop": true
     },
-    "files": [
```
Comment (aj-r) [inline types/lodash/tsconfig.json]: Note: change to `tsconfig.json` so all of the individual function files are included in the lint checks

**Review by aj-r — COMMENTED** [review]
For reviewers: this PR is mainly adding new functions, and I tried to change existing code as little as possible. I've highlighted the notable changes:

**Comment by aj-r** [comment]
@andy-ms any idea what to do about the travis build? Looks like it timed out.

**Comment by aozgaa** [comment]
Hey aj-r, please fix the merge conflicts and we'll fix the build.

**Comment by aj-r** [comment]
Sorry, I was waiting for #24052 to be merged. Working on it.

**Comment by aj-r** [comment]
@aozgaa Done. Unfortunately I needed to make lots of changes to the tests in order to make them complete without running out of memory. The good news is, `npm run lint lodash` now runs in under a minute, which is a vast improvement.

**Comment by andy-ms** [comment]
Test failure is Microsoft/TypeScript#22603

**Comment by aj-r** [comment]
@vsviridov This PR did not include any placeholder support. It's probably possible but will be a little complicated. You might want to create a separate issue for this.

**Comment by vsviridov** [comment]
Created #24602

**Comment by jashworth** [comment]
How do I compose?
```
const f = _flow(_map(x => x), _keyBy('id')); // () => Dictionary<{ ... }> -- no parameters??
f({}); // compiler error: Expected 0 arguments, but got 1.
```

**Comment by aj-r** [comment]
@jashworth try explicitly setting your input parameter types. This works for me:

```ts
import _ from "lodash/fp";

interface TestObject {
    id: string;
}
const f = _.flow(
    _.map((x: TestObject) => x),
    _.keyBy((x: TestObject) => x.id)
);
const keyed = f([{ id: "a" }, { id: "b" }]); // Dictionary<TestObject>
```

**Comment by jashworth** [comment]
Still not working I'm afraid. I'm using TypeScript 2.8.1 and I've npm installed @types/lodash@4.14.106 (is that the correct version?) thanks

image

**Comment by jashworth** [comment]
This feels a bit odd:

```
const p = _flow(_property('key'), key => key !== 'x'); // anything involving flow

const f = _filter(p); // f: (collection: string | null | undefined) => string[]
f([]); // error: argument of type never[] is not assignable to string | null | undefined

// you'll want this most of the time, so you can pass an object or ArrayLike, not a string
const g = _filter<T>(p); // type parameter forces g: (collection: object | ArrayLike<T> | null | undefined) => T[]
g([]); // ok
```

Is there a more natural way to declare g?

**Comment by Aster89** [comment]
Hi, I'm a Vim user that makes use of Vim's plugin YouCompleteMe, which makes in turn use of TSServer for semantic completion and diagnostics for JavaScript.

The thing is, my IDE shows errors when using, for istance, `_.compose`, for which the diagnostic is `Property 'compose' does not exist on type 'LoDashStatic'.`.

[I first asked on YouCompleteMe Gitter chatroom](https://gitter.im/Valloric/YouCompleteMe?at=61b60cb38853d3164044182c); they redirected me to TypeScript, so I filed [a support request there](https://github.com/microsoft/TypeScript/issues/47135); one of their suggestions (with reference to a specific mis-diagnostic) was to ask for help here, which I'm doing.

Furthermore, I also found [a thread about `lodash/fp` on Lodash's repo](https://github.com/lodash/lodash/issues/4326), and there too they suggested to refer to `DefinitelyTyped`.

At the moment I don't have a complete view of what the problem is. I just see that if I use Lodash "the normal way" the IDE doesn't display errors, whereas if I use it "the FP way", some errors are shown (code itself is just fine, to be clear).

Here's my `jsconfig.json`:
```json
{
    "compilerOptions": {
        "checkJs": true,
        "downlevelIteration": true,
        "target": "es6"
    }
}
```
and here is my `package.json`:
```json
{
  "name": "simplemodule",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "kss": "kss --config kss-config.json"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@types/jquery": "^3.5.8",
    "eslint": "^8.7.0",
    "eslint-plugin-lodash-fp": "^2.2.0-a1",
    "kss": "^3.0.1"
  },
  "dependencies": {
    "@types/lodash": "^4.14.178",
    "lodash": "^4.17.21"
  }
}
```
Beside the fact that I've certainly included unrelated pieces in the snippets above, I'm not sure if any of those two is of interest.

I certainly see that among the `dependencies` I have `"@types/lodash": "^4.14.178"`, which makes me think _"shouldn't I have something along the lines of `"@types/lodash/fp": "whate-Ver"`?

**Comment by aj-r** [comment]
Can you share the actual lines of code causing the error? ***@***.***/lodash"
should be enough.

**Comment by aj-r** [comment]
How are you importing? To use fp, you need to import from "lodash/fp", for
example:

import {compose} from "lodash/fp";

compose(...);

**Comment by Aster89** [comment]
> Can you share the actual lines of code causing the error? ***@***.***/lodash" should be enough.

(What is `***@***.***/lodash"` ???)

As I wrote in the previous message, `_.groupBy(_.isEqual, []);` is enough to causing the error.

However, if a more verbose example is needed, in the code below all five occurrences of `compose` are marked in red with the error message `Property 'compose' does not exist on type 'LoDashStatic'. (FixIt)`:
```javascript
  const getPostIdRepAndType = _.compose(
    _.filter('reputation_change'),
    _.map(
      _.compose(
        _.zipObject(['post_type', 'post_id', 'reputation_change']),
        _.over([
          _.compose(_.iteratee('post_type'), _.head),
          _.compose(_.iteratee('post_id'), _.head),
          _.compose(_.sum, _.map(_.iteratee('reputation_change')))
        ])
      )
    ),
    _.values,
    _.groupBy('post_id')
  );
```


> How are you importing? To use fp, you need to import from "lodash/fp", for example:
>
> import {compose} from "lodash/fp";
> compose(...);

And how do I do that in JavaScript? For now I simply have this line in my HTML:
```html
    <script src="https://cdn.jsdelivr.net/g/lodash@4(lodash.min.js+lodash.fp.min.js)"></script>
```
which makes lodash/fp just work.

For the diagnostics, I thought the trick was to configure `package.json` appropriately...

**Comment by aj-r** [comment]
> What is ***@***.***/lodash" ???

Sorry , it looks like github messed up my reply. I meant `@types/lodash` should be enough (there is no package called `@types/lodash/fp` since `lodash/fp` is part of  `@types/lodash`).

It looks like you're loading lodash from a CDN instead of compiling it via NPM. I'm not familiar with that aprroach so I'm not sure what to do. You need to somehow convince TS compiler that `_` has type `_.LodashFp`. Maybe force TS to ambiently load the file [`@types/lodash/fp.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/fp.d.ts)?

**Comment by Aster89** [comment]
@aj-r ,

> It looks like you're loading lodash from a CDN instead of compiling it via NPM.

how can I do otherwise? I mean, probably you are referring to [Custom Builds](https://lodash.com/custom-builds). But would that result in a `.js` minified file that I'd then include in my repository and load in the HTML via `<script>`?

**Comment by aj-r** [comment]
Not a custom build. Normally I include lodash as an NPM dependency, e.g.

```
npm install lodash
```

Then in TypeScript I include it like this:

```
import _ from 'lodash/fp';
```

However this means that your output bundle would include the minified lodash code, so it might slow down your load times compared to the CDN. I'm not sure what the right answer is in this situation.

**Pull request merged**